packages/codecs-core/src/codec.ts

```diff
@@ -59,3 +79,26 @@ export type BaseCodecOptions = {
 export type WrapInCodec<T, U extends T = T> = {
     [P in keyof T]: Codec<T[P], U[P]>;
 };
+
+/**
+ * Get the encoded size of a given value in bytes.
+ */
+export function getEncodedSize<T>(value: T, encoder: Encoder<T>): number {
+    return encoder.fixedSize !== null ? encoder.fixedSize : encoder.variableSize(value);
+}
+
+/**
+ * Use the provided Encoder to encode the given value to a `Uint8Array`.
+ */
+export function encode<T>(value: T, encoder: Encoder<T>): Uint8Array {
```

Comment:
I think my only concern would be that encode and decode are really generic terms for us to be exporting into people's apps, where stuff like getTransactionEncoder is less likely to clash with anything they're doing
They're definitely the right names to use for this though, so not really suggesting any change there!

Reply:
Thanks for the review Callum!

**Regarding the `encode` and `decode` functions**: I agree they are pretty generic. I’d love to keep the `codec.encode` and `codec.decode` API which we could. It just comes with its own set of challenges as codecs will now need to provide 4 functions such that two of them are just delegating to the other two. On the plus side, we no longer have that annoying `[0]` suffix on the main API.

**Regarding offering more complex primitives**: The main issue here is that the current `encode` and `decode` logic are not symmetrical. Meaning there are things we can do with `decode` that we cannot do with `encode` because we do not have access to the entire buffer. I do have some examples of that. For instance, imagine an `offsetCodec(codec, relativeOffset)` helper that, given a codec, shifts its position on the buffer by the given relative offset. We can do that right now with `Decoders` but it’s impossible to do with `Encoders` because they use a different mechanism that rely on linear serialisation only. An example of why we would we need something like an `offsetCodec` helper is any array such that its size does not immediately prefix the items.

**Regarding performance improvements**: The decoding logic should stay pretty much the same because it already uses a single buffer but the encoding logic should drastically use less instances of `Uint8Array` which I believe will be a big performance boost.